packages/generator/test/generator.spec.ts

```diff
@@ -1,46 +1,123 @@
 import { generateProject } from '../src';
 import { createOptions } from './test-util/create-generator-options';
 
 describe('generator', () => {
-  it('generates expected files', async () => {
-    const project = await generateProject(
-      createOptions({
-        inputDir: '../../test-resources/service-specs/v2/API_TEST_SRV',
-        useSwagger: false
-      })
-    );
-    const files = project!.getSourceFiles();
-
-    expect(files.length).toBe(28);
-
-    const testEntityFile = files.find(
-      file => file.getBaseName() === 'TestEntity.ts'
-    );
-
-    expect(testEntityFile).toBeDefined();
-    expect(testEntityFile!.getClasses().length).toBe(1);
-    expect(testEntityFile!.getInterfaces().length).toBe(2);
-    expect(testEntityFile!.getNamespaces().length).toBe(1);
-
-    const entityClass = testEntityFile!.getClass('TestEntity');
-    expect(entityClass!.getProperties().length).toBe(24);
-
-    const properties = entityClass!.getProperties();
-    const staticProperties = [
-      properties.find(p => p.getName() === '_entityName')!,
-      properties.find(p => p.getName() === '_serviceName')!,
-      properties.find(p => p.getName() === '_defaultServicePath')!
-    ];
-
-    expect(staticProperties.map(p => p.isStatic())).toEqual([true, true, true]);
-    expect(staticProperties.map(p => p.getInitializer()!.getText())).toEqual([
-      "'A_TestEntity'",
-      "'API_TEST_SRV'",
-      "'/sap/opu/odata/sap/API_TEST_SRV'"
-    ]);
-
-    const entityNamespace = testEntityFile!.getNamespace('TestEntity');
-    expect(entityNamespace!.getVariableDeclarations().length).toBe(25);
+  let files: SourceFile[];
+
+  describe('v2', () => {
+    beforeAll(async () => {
+      files = await getGeneratedFiles('v2');
+    });
+
+    it('generates expected number of files', () => {
+      expect(files.length).toBe(28);
+    });
+
+    it('generates TestEntity.ts file', () => {
+      const testEntityFile = files.find(
+        file => file.getBaseName() === 'TestEntity.ts'
+      );
+
+      expect(testEntityFile).toBeDefined();
+      expect(testEntityFile!.getClasses().length).toBe(1);
+      expect(testEntityFile!.getInterfaces().length).toBe(2);
+      expect(testEntityFile!.getNamespaces().length).toBe(1);
+
+      const entityClass = testEntityFile!.getClass('TestEntity');
+      expect(entityClass!.getProperties().length).toBe(24);
+
+      checkStaticProperties(entityClass!);
+
+      const entityNamespace = testEntityFile!.getNamespace('TestEntity');
+      expect(entityNamespace!.getVariableDeclarations().length).toBe(25);
+    });
+
+    it('generates function-imports.ts file', () => {
+      const functionImports = getFunctionImportDeclarations(files);
+      expect(functionImports.length).toBe(12);
+    });
+  });
+
+  describe('v4', () => {
+    beforeAll(async () => {
+      files = await getGeneratedFiles('v4');
+    });
+
+    it('generates expected number of files', () => {
+      expect(files.length).toBe(29);
```

Comment:
There is a similar test above about v2.
Could you please compare these numbers and explain a bit regarding the difference?
In this case, what's the additional file?

Reply:
`TestComplexBaseType` - there are now base types in the edmx definitions for entities and complex types.